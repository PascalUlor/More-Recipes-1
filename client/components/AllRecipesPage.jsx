/* eslint-disable */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Pagination from 'rc-pagination';
import Spinner from 'react-md-spinner';
import NavBar from './NavBar.jsx';
import PageHeader from './allRecipesPage/PageHeader.jsx';
import TextFieldGroup from '../common/TextFieldGroup.jsx';
import AllRecipesList from './allRecipesPage/AllRecipesList.jsx';
import PopularRecipesList from './allRecipesPage/PopularRecipesList.jsx';
import Footer from './Footer.jsx';
import fetchPopularRecipesRequest from '../actions/actionCreators/popularRecipesActions';
import {
  fetchAllRecipesRequest
} from '../actions/actionCreators/getAllRecipesActions';


/**
 * @description HOC for all recipes component
 *
 * @class AllRecipesPage
 *
 * @extends Component
 */
export class AllRecipesPage extends Component {
  /**
   * @description creates an instance of AllRecipesPage
   *
   * @constructor
   *
   * @param { props } props - contains all recipes component properties
   *
   */
  constructor(props) {
    super(props);
    this.state = {
      search: ''
    };
    this.handlePageChange = this.handlePageChange.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
  }
  /**
   *; @description handles fetching all and popular recipes
   *
   * @method componentDidMount
   *
   * @returns { * } null
   */
  componentDidMount() {
    const { search } = this.state;
    this.props.fetchAllRecipes(1, search);
    this.props.fetchPopularRecipes();
  }
  /**
   * @description handles search value onchange event
   * @method handleChange
   *
   * @param { object } event -
   *
   * @returns {*} null
   */
  handleChange(event) {
    this.setState({
      search: event.target.value
    });
  }
  /**
   * @description handles search value onchange event
   * @method handleSearch
   *
   * @param { object } event -
   *
   * @returns {*} null
   */
  handleSearch() {
    const { search } = this.state;
    if (search.length >= 3) {
      this.props.fetchAllRecipes(1, search);
    } else if (search.length === 0) {
      this.props.fetchAllRecipes(1, search);
    }
  }
  /**
   * @description handles page navigation
   * @method handlePageChange
   *
   * @param {number} page - page number to fetch
   *
   * @returns {*} null
   */
  handlePageChange(page) {
    const { search } = this.state;
    this.props.fetchAllRecipes(page, search);
  }
  /**
   * @description renders all and popular recipes
   *
   * @returns { jsx } jsx - renders all and popular recipes components
   */
  render() {
    const {
        currentPage, limit, numberOfRecipes
      } = this.props.paginationDetails,
      {
        isAllFetching, isPopularFetching, errorMessage,
        fetchedAllRecipes, fetchedPopularRecipes
      } = this.props,
      { search } = this.state;

    let displayRecipes;
    if (fetchedAllRecipes.length < 1 && fetchedPopularRecipes < '/') {
      displayRecipes = (
        <div className="not-found lead p-3 mt-3 mt-5">
          <i className=
            "fa fa-exclamation-triangle fa-3x pb-3 text-warning d-block">
          </i>
          {errorMessage}
        </div>);
    } else {
      displayRecipes = (
        <div className="row">
          <div className="col-sm-7 col-md-9 col-lg-9">
            <AllRecipesList
              allRecipes={fetchedAllRecipes}
              errorMessage={errorMessage}/>
            {(numberOfRecipes > 6 && typeof numberOfRecipes !== 'undefined') &&
          <Pagination
            onChange={this.handlePageChange}
            current={currentPage}
            pageSize={limit || 0}
            total={numberOfRecipes}
            className="pagination"
          />}
          </div>
          <div className="col-sm-5 col-md-3 col-lg-3">
            <h6 className="card-title page-text text-center mt-2">Popular Recipes
            </h6>
            <PopularRecipesList popularRecipes={fetchedPopularRecipes}/>
          </div>
        </div>);
    }

    return (
      <div className="bg-faded">
        <div className="site-wrapper">
          <NavBar/>
          <div className="container">
            <main className="main-wrapper pt-3">
              <PageHeader/>
              <form
                className="col-10 offset-1 col-sm-8 col-lg-6 offset-lg-3"
                onKeyUp={this.handleSearch}>
                <TextFieldGroup
                  font="fa fa-search" name='search'
                  value={search} onChange={this.handleChange}
                  placeholder='search by recipe name e.g fried rice'/>
              </form>
              {(isAllFetching || isPopularFetching) ?
                <div className="text-center">
                  <Spinner size={50} className="mt-5 mb-5"/>
                </div>
                :
                <div>
                  {displayRecipes}
                </div>
              }
            </main>
          </div>
        </div>
        <Footer/>
      </div>
    );
  }
}

AllRecipesPage.propTypes = {
  isPopularFetching: PropTypes.bool.isRequired,
  isAllFetching: PropTypes.bool.isRequired,
  fetchedPopularRecipes: PropTypes.array.isRequired,
  fetchedAllRecipes: PropTypes.array.isRequired,
  paginationDetails: PropTypes.shape().isRequired,
  fetchPopularRecipes: PropTypes.func.isRequired,
  fetchAllRecipes: PropTypes.func.isRequired,
  errorMessage: PropTypes.string.isRequired
};
/**
 * @description maps redux state to props
 *
 * @param { object } state - holds all recipes state
 *
 * @return { object } props - returns mapped props from state
 */
const mapStateToProps = state => ({
  isPopularFetching: state.popularRecipes.isPopularRecipesFetching,
  isAllFetching: state.allRecipes.isAllRecipesFetching,
  fetchedPopularRecipes: state.popularRecipes.fetchedPopularRecipes,
  fetchedAllRecipes: state.allRecipes.fetchedAllRecipes,
  paginationDetails: state.allRecipes.paginationDetails,
  errorMessage: state.allRecipes.allRecipesError
});
/**
 * @description maps action dispatch to props
 *
 * @param { object } dispatch - holds dispatchable actions
 *
 * @return { object } props - returns mapped props from dispatch action
 */
const mapDispatchToProps = dispatch => ({
  fetchPopularRecipes: () => dispatch(fetchPopularRecipesRequest()),
  fetchAllRecipes: (page, search) => dispatch(fetchAllRecipesRequest(page, search))
});

export default connect(mapStateToProps, mapDispatchToProps)(AllRecipesPage);
