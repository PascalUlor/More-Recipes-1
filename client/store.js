import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import indexReducer from './reducers/index';


const env = process.env.NODE_ENV || development;

let middleware = compose(applyMiddleware(thunk,
    reduxImmutableStateInvariant()), window.devToolsExtension ?
  window.devToolsExtension() : f => f);

if (env === 'production') {
  middleware = applyMiddleware(thunk);
}

const store = createStore(
  indexReducer,
  middleware
);

export default store;
