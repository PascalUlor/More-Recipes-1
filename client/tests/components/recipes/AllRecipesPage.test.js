/* eslint-disable */
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import store from '../../../store';
import AllRecipesPage from '../../../components/AllRecipesPage.jsx';


const props = {};

const mountWrapper = mount((
  <Provider store={store}>
    <Router><AllRecipesPage {...props}/></Router>
  </Provider>
));
describe('AllRecipesPage component', () => {
  it('should render correctly', () => {
    expect(toJson(mountWrapper)).toMatchSnapshot();
  });

  it('should call onChange() event', () => {
    mountWrapper.find('TextFieldGroup').find('input')
      .simulate('change', {
        target: {
          name: 'search',
          value: 'rice',
        },
      });
    expect(mountWrapper.find('AllRecipesPage')
      .instance().state.search).toEqual('rice');
  });

  it('should call handleSearch()', () => {
    mountWrapper.find('form').simulate('keyup');
    expect(mountWrapper.find('AllRecipesPage').instance().state.search)
      .toEqual('rice');
    mountWrapper.find('AllRecipesPage').instance().setState({
      search: '',
    });
    mountWrapper.find('form').simulate('keyup');
    expect(mountWrapper.find('AllRecipesPage').instance().state.search)
      .toEqual('');
  });
});
