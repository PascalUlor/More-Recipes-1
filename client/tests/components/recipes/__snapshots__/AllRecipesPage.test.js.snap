// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllRecipesPage component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <BrowserRouter>
      <AllRecipesPage
        errorMessage=""
        fetchAllRecipes={[MockFunction]}
        fetchPopularRecipes={[MockFunction]}
        fetchedAllRecipes={Array []}
        fetchedPopularRecipes={Array []}
        isAllFetching={false}
        isPopularFetching={false}
        paginationDetails={Object {}}
      />
    </BrowserRouter>
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <AllRecipesPage
        errorMessage=""
        fetchAllRecipes={[MockFunction]}
        fetchPopularRecipes={[MockFunction]}
        fetchedAllRecipes={Array []}
        fetchedPopularRecipes={Array []}
        isAllFetching={false}
        isPopularFetching={false}
        paginationDetails={Object {}}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "errorMessage": "",
        "fetchAllRecipes": [MockFunction],
        "fetchPopularRecipes": [MockFunction],
        "fetchedAllRecipes": Array [],
        "fetchedPopularRecipes": Array [],
        "isAllFetching": false,
        "isPopularFetching": false,
        "paginationDetails": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <AllRecipesPage
          errorMessage=""
          fetchAllRecipes={[MockFunction]}
          fetchPopularRecipes={[MockFunction]}
          fetchedAllRecipes={Array []}
          fetchedPopularRecipes={Array []}
          isAllFetching={false}
          isPopularFetching={false}
          paginationDetails={Object {}}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "errorMessage": "",
          "fetchAllRecipes": [MockFunction],
          "fetchPopularRecipes": [MockFunction],
          "fetchedAllRecipes": Array [],
          "fetchedPopularRecipes": Array [],
          "isAllFetching": false,
          "isPopularFetching": false,
          "paginationDetails": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
